### PR TITLE
[RFC] vim-patch:8.0.1494: no autocmd triggered in Insert mode with visible …

### LINKED ISSUE #1691: TextChangedI when pumvisible
Currently `TextChangedI` isn't triggered when the popup menu is visible. This means all autocomplete plugins have to use timer based hacks.
Would it be possible to have a new event something like `TextChangedIEx` which is triggered regardless of popup menu being visible or not? This way I could use timer as fallback instead in [asyncomplete.vim](https://github.com/prabirshrestha/asyncomplete.vim). All other async completion plugins - [completor.vim](https://github.com/maralla/completor.vim), [deoplete](https://github.com/Shougo/deoplete.nvim), [ncm](https://github.com/roxma/nvim-completion-manager) also seems to use timer based hacks.
I'm also writing a [language server plugin for vim](https://github.com/prabirshrestha/vim-lsp) and see a use case here to detect text document changes.

https://groups.google.com/forum/#!topic/vim_dev/DrDYKWUeHVY

### PR DESCRIPTION
**Pull request by Shougo**
…popup menu

Problem:    No autocmd triggered in Insert mode with visible popup menu.
Solution:   Add TextChangedP. (Prabir Shrestha, Christian Brabandt,
            closes vim/vim#2372, closes vim/vim#1691)
            Fix that the TextChanged autocommands are not always triggered
            when sourcing a script.

https://github.com/vim/vim/commit/5a093437199001a0d60d8e18e2b9539b99a7757c

### CODE REVIEW

**Comment by Shougo** [comment]
I have tested the feature in deoplete and it works.

File: src/nvim/edit.c [hunk 1]
```diff
@@ -1375,13 +1375,19 @@ ins_redraw (
...
-    last_changedtick_buf = curbuf;
-    last_changedtick = curbuf->b_changedtick;
+      curbuf->b_last_changedtick = curbuf->b_changedtick;
+  }
+
+  // Trigger TextChangedP if b_changedtick differs. When the popupmenu closes
+  // TextChangedI will need to trigger for backwards compatibility, thus use
+  // different b_last_changedtick* variables.
+  if (ready && curbuf->b_last_changedtick_pum != curbuf->b_changedtick
+          && pum_visible()) {
```
Comment (justinmk) [inline src/nvim/edit.c]: why was the check for `has_textchangedP()` not included?

Please don't silently make changes to ported patches without explaining why. It makes it really difficult to trust that the patches are being ported correctly.

Comment (Shougo) [reply src/nvim/edit.c]: `has_textchangedP()` implementation uses `autopat` feature.
It is not included in neovim.
And it just checks auto command is defined.

Comment (bfredl) [reply src/nvim/edit.c]: We have generic `has_event`, it should be used instead.

Comment (Shougo) [reply src/nvim/edit.c]: OK.   I will fix it.

**Comment by justinmk** [comment]
```
Caught exception in Test_ChangedP(): Vim(call):E117: Unknown function: test_override @ function RunTheTest[13]..Test_ChangedP, line 3
TEST FAILURE
```

We don't support test_override. That's starting to become a problem, Vim loves to use that in tests.

**Comment by justinmk** [comment]
It should be easy to convert the test to `test/functional/viml/completion_spec.lua`. Use one of the tests there as an example.

**Comment by prabirshrestha** [comment]
👍  Looking forward to this getting merged in neovim. I got asyncomplete.vim working with `TextChangedP` at https://github.com/prabirshrestha/asyncomplete.vim/pull/49

**Comment by Shougo** [comment]
I  have fixed the conflicts.

>  throw 'skipped: Nvim does not support test_override()'

I have added the skip.

> It should be easy to convert the test to test/functional/viml/completion_spec.lua

Sorry, I don't have time to convert the tests.

**Comment by Shougo** [comment]
I need to port the other patches.

**Comment by blueyed** [comment]
@justinmk 

> We don't support test_override. That's starting to become a problem, Vim loves to use that in tests.

What's the reason for this?

**Comment by justinmk** [comment]
test_override is a hack because vim doesn't have screen tests. It silently changes internal behavior.

I wouldn't object to implemeting it for tests _only_, I.e. we would disable it in releases.

**Comment by jamessan** [comment]
> I wouldn't object to implemeting it for tests only, I.e. we would disable it in releases.

And disable any tests using it, since I would expect the tests to pass when I build the Debian package.

**Comment by justinmk** [comment]
@jamessan Yes, tests will need to check `has('*test_override')`.

File: src/nvim/testdir/test_autocmd.vim [hunk 2]
```diff
@@ -823,6 +823,7 @@ func Test_Cmdline()
   au! CmdlineLeave
 endfunc
 
+<<<<<<< HEAD
```
Comment (prabirshrestha) [inline src/nvim/testdir/test_autocmd.vim]: seems like a bad merge here.

Comment (Shougo) [reply src/nvim/testdir/test_autocmd.vim]: I will fix it later.

**Comment by Shougo** [comment]
Fixed.

**Comment by Cosson2017** [comment]
This patch is awesome!

**Comment by Shougo** [comment]
https://github.com/SpaceVim/SpaceVim/issues/1558

TextChangedP is called when `<C-n>/<C-p>`...
It is hard to use in auto completion plugin to trigger the completion.

**Comment by chemzqm** [comment]
Any progress for this PR? I'm looking forward to it.

**Comment by chemzqm** [comment]
@justinmk Where to add the line `  throw 'skipped: Nvim does not support test_override()'`  at the end of vim test code ?

**Pull request closed without merging**